File: shogi/src/shogi/kifu/csa/csa.go
package csa

import (
	"fmt"
	"strings"
	. "shogi"
	"shogi/kifu"
)

var registered = kifu.AddLoader( kifu.Loader{ []string{".csa"}, Parse } )

func atoi( c uint8 ) int {
	return int(c) - '0'
}

func Parse( str string ) ( kifu.Kifu, error ) {
	coms := []Command{}
	b := new(Board)
	
	src := strings.Split( str, "\n" )
	version := 0
	for _,line := range src {
		if len(line) == 0 { continue }
		switch line[0] {
		case '\'':
			// コメント
		case 'V':
			version = atoi(line[1])
		case 'N':
			// 対戦者名
		case '$':
			// 付加情報
		case 'T':
			// 時間情報
		case '%':
			// 特殊コマンド
		case 'P':
			switch line[1] {
			case '1','2','3','4','5','6','7','8','9':
				y := int(line[1]) - '0'
				for i := 0; i <= 9; i++ {
					if len(line) <= 2+i*3 { break }
					if line[2+i*3] == '-' || line[2+i*3] == '+' {
						var player Player
						if line[2+i*3] == '+' { player = Sente }else{ player = Gote }
						koma := KomaFromString( line[2+i*3+1:2+i*3+3] )
						b.SetCell( MakePos(9-i,y), MakeKoma( koma, player ) )
					}
				}
				_ = y
			case 'I':
				b.Init()
			case '+','-':
			default:
				println( line )
			}
		case '+','-':
			if len(line) == 1 {
				// 手番設定
				if( line == "+" ){
					b.Teban = Sente
				}else{
					b.Teban = Gote
				}
			}else{
				com, err := ParseCommand( line )
				if err != nil { fmt.Println( "error: "+line ) }
				coms = append( coms, com )
			}
		default:
			println( line )
		}
	}
	_ = version

	r := new(kifu.KifuBase)
	r.ACommands = coms
	r.ABoard = b
	return r, nil
}

File: cshogi/csa.cpp
#include "csa.h"
#include <boost/algorithm/string.hpp>
#include <iostream>

namespace csa {
	
Kifu* parse( istream &in )
{
	Kifu *kifu = new Kifu();
	Board *board = new Board();
	vector<Command> coms;

	string line;
	while( in && getline( in, line ) ){
		if( line.length() == 0 ) continue;
		switch( line[0] ){
		case '\'': // コメント
		case 'N': // 対戦者名
		case '$': // 付加情報
		case 'T': // 時間情報
		case '%': // 特殊コマンド
		case 'V': // バージョン
			break;
		case 'P':
			switch( line[1] ){
			case '1': case '2': case '3': case '4': case '5': case '6': case '7': case '8': case '9':
				{
					int y = line[1] - '0';
					for( int i = 0; i <= BOARD_SIZE; i++ ){
						if( line.length() <= 2+i*3 ) break;
						if( line[2+i*3] == '-' || line[2+i*3] == '+' ){
							Player player = (line[2+i*3] == '+' )?SENTE:GOTE;
							KomaKind koma = KomaKindFromString( line.substr(2+i*3+1,2).c_str() );
							board->SetCell( Pos(9-i,y), Koma( koma, player ) );
						}
					}
				}
				break;
			case 'I':
				board->Init();
				break;
			case '+':
			case '-':
				{
					Player player = (line[1] == '+')?SENTE:GOTE;
					for( int i=0; i<(line.length()-2)/4; ++i ){
						Pos pos(line[i*4+2] - '0', line[i*4+3] - '0');
						KomaKind koma = KomaKindFromString( line.substr(i*4+4,2).c_str() );
						if( pos == KOMADAI ){
							board->hand[player][koma] += 1;
						}else{
							board->SetCell( pos, Koma( koma, player ) );
						}
					}
				}
				break;
			default:
				cout << line << endl;
				break;
			}
			break;
		case '+':
		case '-':
			if( line.length() == 1 ){
				board->curPlayer = (line[0] == '+')?SENTE:GOTE;
			}else{
				coms.push_back( Command( line ) );
			}
			break;
		default:
			cout << line << endl;
		}
	}

	return new Kifu( board, coms );
}
	
}
